### PR TITLE
[MI-2450] Added testcases and Makefile command to generate mocks

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
1. TestGetSubscriptionFIlterPossibleValues
2. TestCheckIfUserIsProjectAdmin
3. TestGetBuildDetails
4. TestPostBuildDetailsPreview
5. TestHandleDeleteAllSubscriptions
6. TestHandleGetSubscriptionFilterPossibleValues
7. command.go

### CODE REVIEW

File: server/plugin/api_test.go [hunk 1]
```diff
@@ -295,6 +321,65 @@ func TestHandleLink(t *testing.T) {
...
+			userID:                "mockMattermostUserID",
+			projectID:             "mockProjectID",
+			statusCode:            http.StatusInternalServerError,
+			getAllSubscriptionErr: errors.New("mockError"),
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("LogError", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string"))
+
+			if testCase.statusCode == http.StatusOK || testCase.statusCode == http.StatusInternalServerError {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:368]: i think in the above test cases  there are only 2 status codes and both are in this condition so there is no need of this condition

File: server/plugin/api_test.go [hunk 2]
```diff
@@ -295,6 +321,65 @@ func TestHandleLink(t *testing.T) {
...
+	mockAPI := &plugintest.API{}
+	mockCtrl := gomock.NewController(t)
+	mockedClient := mocks.NewMockClient(mockCtrl)
+	mockedStore := mocks.NewMockKVStore(mockCtrl)
+	p := setupMockPlugin(mockAPI, mockedStore, mockedClient)
+	for _, testCase := range []struct {
+		description           string
+		userID                string
+		projectID             string
+		err                   error
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:335]: you have not used this err anywhere

File: server/plugin/api_test.go [hunk 3]
```diff
@@ -295,6 +321,65 @@ func TestHandleLink(t *testing.T) {
...
+		statusCode            int
+		getAllSubscriptionErr error
+		subscriptionList      []*serializers.SubscriptionDetails
+		subscription          *serializers.SubscriptionDetails
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:339]: same

File: server/plugin/api_test.go [hunk 4]
```diff
@@ -295,6 +321,65 @@ func TestHandleLink(t *testing.T) {
...
+			if testCase.statusCode == http.StatusOK || testCase.statusCode == http.StatusInternalServerError {
+				mockedStore.EXPECT().GetAllSubscriptions(testCase.userID).Return(testCase.subscriptionList, testCase.getAllSubscriptionErr)
+			}
+
+			if testCase.statusCode == http.StatusOK {
+				mockedClient.EXPECT().DeleteSubscription(gomock.Any(), gomock.Any(), gomock.Any()).Return(testCase.statusCode, testCase.err)
+				mockedStore.EXPECT().DeleteSubscription(gomock.Any()).Return(nil)
+			}
+
+			statusCode, _ := p.handleDeleteAllSubscriptions(testCase.userID, testCase.projectID)
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:377]: check whether the error is nil or not

File: server/plugin/api_test.go [hunk 5]
```diff
@@ -1112,3 +1197,102 @@ func TestHandleGetGitRepositoryBranches(t *testing.T) {
...
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("LogError", testutils.GetMockArgumentsWithType("string", 3)...)
+
+			if testCase.statusCode == http.StatusOK || testCase.statusCode == http.StatusInternalServerError {
+				mockedClient.EXPECT().GetSubscriptionFilterPossibleValues(gomock.Any(), gomock.Any()).Return(testCase.getGitRepositoryBranchesResponse, testCase.statusCode, testCase.getSubscriptionFilterPossibleValuesErr)
+			}
+
+			req := httptest.NewRequest(http.MethodGet, "/mockPath", bytes.NewBufferString(testCase.body))
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1276]: not able to understand this you are not calling the actual API and what is the need for this then ?

File: server/plugin/client_test.go [hunk 6]
```diff
@@ -392,6 +431,87 @@ func TestGetGitRepositoryBranches(t *testing.T) {
...
+			err:         errors.New("mock-error"),
+			statusCode:  http.StatusInternalServerError,
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			monkey.PatchInstanceMethod(reflect.TypeOf(&client{}), "Call", func(_ *client, basePath, method, path, contentType, mattermostUserID string, inBody io.Reader, out interface{}, formValues url.Values) (responseData []byte, statusCode int, err error) {
+				return nil, testCase.statusCode, testCase.err
+			})
+
+			_, statusCode, err := p.Client.GetSubscriptionFilterPossibleValues(&testCase.request, "mockMattermostUSerID")
```
Comment (Kshitij-Katiyar) [inline server/plugin/client_test.go:502]:
```suggestion
			_, statusCode, err := p.Client.GetSubscriptionFilterPossibleValues(&testCase.request, "mockMattermostUserID")
```

File: server/plugin/command_test.go [hunk 7]
```diff
@@ -63,21 +75,154 @@ func TestExecuteCommand(t *testing.T) {
...
+		},
+		{
+			description:      "ExecuteCommand: boards list subscription command",
+			isConnected:      true,
+			commandArgs:      &model.CommandArgs{Command: "/azuredevops boards subscription list me"},
+			isListCommand:    true,
+			ephemeralMessage: "mockSubscriptionList",
+		},
+		{
+			description:             "ExecuteCommand: failed to get all subscriptions",
```
Comment (Kshitij-Katiyar) [inline server/plugin/command_test.go:132]: Check this everywhere
```suggestion
			description:             "ExecuteCommand: failed to get all the subscriptions",
```

File: server/plugin/command_test.go [hunk 8]
```diff
@@ -104,3 +268,81 @@ func TestExecuteCommand(t *testing.T) {
...
+	p.Client = mockedClient
+	for _, testCase := range []struct {
+		description      string
+		commandArgs      *model.CommandArgs
+		ephemeralMessage string
+		isConnected      bool
+		patchAPICalls    func()
+		expectedError    *model.AppError
+	}{
+		// {
```
Comment (Kshitij-Katiyar) [inline server/plugin/command_test.go:290]: need for this ?

File: server/plugin/command_test.go [hunk 9]
```diff
@@ -104,3 +268,81 @@ func TestExecuteCommand(t *testing.T) {
...
+			})
+
+			monkey.Patch(azureDevopsDeleteCommand, func(_ *Plugin, _ *plugin.Context, _ *model.CommandArgs, _ string, _ ...string) (*model.CommandResponse, *model.AppError) {
+				return &model.CommandResponse{}, testCase.expectedError
+			})
+
+			monkey.PatchInstanceMethod(reflect.TypeOf(&p), "ParseSubscriptionsToCommandResponse", func(_ *Plugin, _ []*serializers.SubscriptionDetails, _, _, _, _, _ string) string {
+				return "mockSubscriptionList"
+			})
+			// monkey.Patch(azureDevopsListSubscriptionsCommand, func(_ *Plugin, _ *plugin.Context, _ *model.CommandArgs, _ string, _ ...string) (*model.CommandResponse, *model.AppError) {
```
Comment (Kshitij-Katiyar) [inline server/plugin/command_test.go:323]: ?

File: server/plugin/command_test.go [hunk 10]
```diff
@@ -63,21 +75,154 @@ func TestExecuteCommand(t *testing.T) {
...
+		},
+		{
+			description: "ExecuteCommand: pipelines add subscription command",
+			isConnected: true,
+			commandArgs: &model.CommandArgs{Command: "/azuredevops pipelines subscription add"},
+		},
+		{
+			description:      "ExecuteCommand: pipelines command when user is not connected",
+			commandArgs:      &model.CommandArgs{Command: "/azuredevops pipelines wrong [title] [description]"},
+			ephemeralMessage: fmt.Sprintf(constants.ConnectAccountFirst, fmt.Sprintf(constants.ConnectAccount, p.GetPluginURLPath(), constants.PathOAuthConnect)),
```
Comment (ayusht2810) [inline server/plugin/command_test.go:182]: this test case is repeated

File: server/plugin/command_test.go [hunk 11]
```diff
@@ -104,3 +268,81 @@ func TestExecuteCommand(t *testing.T) {
 		})
 	}
 }
+
+func TestExecuteCommasnd(t *testing.T) {
```
Comment (ayusht2810) [inline server/plugin/command_test.go:272]:
```suggestion
func TestExecuteCommand(t *testing.T) {
```

File: server/plugin/command_test.go [hunk 12]
```diff
@@ -104,3 +268,81 @@ func TestExecuteCommand(t *testing.T) {
...
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("SendEphemeralPost", mock.AnythingOfType("string"), mock.AnythingOfType("*model.Post")).Run(func(args mock.Arguments) {
+				post := args.Get(1).(*model.Post)
+				assert.Equal(t, testCase.ephemeralMessage, post.Message)
+			}).Once().Return(&model.Post{})
+
+			mockAPI.On("GetBundlePath").Return("/test-path", nil)
+			mockAPI.On("LogError", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string"))
```
Comment (ayusht2810) [inline server/plugin/command_test.go:309]: use GetMockArgumentsWithType

File: server/plugin/command_test.go [hunk 13]
```diff
@@ -104,3 +268,81 @@ func TestExecuteCommand(t *testing.T) {
...
+			mockAPI.On("PublishWebSocketEvent", mock.AnythingOfType("string"), mock.Anything, mock.AnythingOfType("*model.WebsocketBroadcast")).Return(nil)
```
Comment (ayusht2810) [inline server/plugin/command_test.go:310]:
```suggestion
			mockAPI.On("PublishWebSocketEvent", mock.AnythingOfType("string"), mock.Anything, mock.AnythingOfType("*model.WebsocketBroadcast")).Return()
```

File: server/plugin/api_test.go (hunk 5 above)
Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:1125]: We are adding the `Mattermost-User-ID` header in this req and passing it to the function call

File: server/plugin/api_test.go [hunk 14]
```diff
@@ -294,6 +321,91 @@ func TestHandleLink(t *testing.T) {
...
+					SubscriptionID:   "mockSubscriptionID",
+				},
+			},
+		},
+		{
+			description:            "HandleDeleteAllSubscriptions: GetAllSubscriptions gives error",
+			userID:                 "mockMattermostUserID",
+			projectID:              "mockProjectID",
+			statusCode:             http.StatusInternalServerError,
+			getAllSubscriptionsErr: errors.New("mockError"),
```
Comment (ayusht2810) [inline server/plugin/api_test.go:362]: change error

File: server/plugin/api_test.go [hunk 15]
```diff
@@ -294,6 +321,91 @@ func TestHandleLink(t *testing.T) {
...
+				{
+					MattermostUserID: "mockMattermostUserID",
+					ProjectID:        "mockProjectID",
+					OrganizationName: "mockOrganization",
+					EventType:        "mockEventType",
+					ChannelID:        "mockChannelID",
+					SubscriptionID:   "mockSubscriptionID",
+				},
+			},
+			err:                  errors.New("mockError"),
```
Comment (ayusht2810) [inline server/plugin/api_test.go:380]: same

File: server/plugin/api_test.go [hunk 16]
```diff
@@ -294,6 +321,91 @@ func TestHandleLink(t *testing.T) {
...
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("LogError", testutils.GetMockArgumentsWithType("string", 3)...)
+
+			mockedStore.EXPECT().GetAllSubscriptions(testCase.userID).Return(testCase.subscriptionList, testCase.getAllSubscriptionsErr)
+
+			if testCase.getAllSubscriptionsErr == nil {
+				mockedClient.EXPECT().DeleteSubscription(gomock.Any(), gomock.Any(), gomock.Any()).Return(testCase.statusCode, testCase.err)
+			}
+
+			if testCase.getAllSubscriptionsErr == nil && testCase.err == nil {
```
Comment (ayusht2810) [inline server/plugin/api_test.go:393]: we can combine this with the above if

File: server/plugin/api_test.go [hunk 17]
```diff
@@ -266,16 +269,40 @@ func TestHandleLink(t *testing.T) {
...
 
+			monkey.PatchInstanceMethod(reflect.TypeOf(p), "IsProjectLinked", func(*Plugin, []serializers.ProjectDetails, serializers.ProjectDetails) (*serializers.ProjectDetails, bool) {
+				return &serializers.ProjectDetails{}, testCase.isProjectLinked
+			})
+
 			if testCase.statusCode == http.StatusOK {
+				mockedStore.EXPECT().GetAllProjects("mockMattermostUserID").Return(testCase.projectList, nil)
+			}
+
+			if testCase.statusCode == http.StatusOK && !testCase.isProjectLinked {
```
Comment (ayusht2810) [inline server/plugin/api_test.go:303]: same

File: server/plugin/client_test.go [hunk 18]
```diff
@@ -172,6 +173,46 @@ func TestGetPullRequest(t *testing.T) {
...
+	p := setupTestPlugin(mockAPI)
+	for _, testCase := range []struct {
+		description          string
+		err                  error
+		statusCode           int
+		expectedErrorMessage string
+	}{
+		{
+			description: "GetBuildDetails: valid",
+			err:         nil,
```
Comment (ayusht2810) [inline server/plugin/client_test.go:188]:
```suggestion
```

File: server/plugin/client_test.go [hunk 19]
```diff
@@ -317,6 +358,88 @@ func TestCall(t *testing.T) {
...
+		err         error
+		statusCode  int
+	}{
+		{
+			description: "CheckIfUserIsProjectAdmin: valid",
+			statusCode:  http.StatusOK,
+		},
+		{
+			description: "CheckIfUserIsProjectAdmin: with error",
+			err:         errors.New("mock-error"),
```
Comment (ayusht2810) [inline server/plugin/client_test.go:375]: change error

File: server/plugin/client_test.go [hunk 20]
```diff
@@ -317,6 +358,88 @@ func TestCall(t *testing.T) {
...
+			request: &serializers.GetSubscriptionFilterPossibleValuesRequestPayload{
+				Filters:      []string{"mockFIlter1", "mockFilter2"},
+				EventType:    "mockEventType",
+				ProjectID:    "mockProjectID",
+				RepositoryID: "mockRepositoryID",
+			},
+		},
+		{
+			description:          "GetSubscriptionFilterPossibleValues: with error",
+			err:                  errors.New("mock-error"),
```
Comment (ayusht2810) [inline server/plugin/client_test.go:419]: same

File: server/plugin/client_test.go [hunk 21]
```diff
@@ -317,6 +358,88 @@ func TestCall(t *testing.T) {
...
+		err                  error
+		statusCode           int
+		request              *serializers.GetSubscriptionFilterPossibleValuesRequestPayload
+		expectedErrorMessage string
+	}{
+		{
+			description: "GetSubscriptionFilterPossibleValues: valid",
+			statusCode:  http.StatusOK,
+			request: &serializers.GetSubscriptionFilterPossibleValuesRequestPayload{
+				Filters:      []string{"mockFIlter1", "mockFilter2"},
```
Comment (ayusht2810) [inline server/plugin/client_test.go:411]:
```suggestion
				Filters:      []string{"mockFilter1", "mockFilter2"},
```

File: server/plugin/command_test.go [hunk 22]
```diff
@@ -15,21 +17,32 @@ import (
...
 func TestExecuteCommand(t *testing.T) {
+	monkey.UnpatchAll()
 	p := Plugin{}
 	mockAPI := &plugintest.API{}
 	mockCtrl := gomock.NewController(t)
 	mockedStore := mocks.NewMockKVStore(mockCtrl)
+	mockedClient := mocks.NewMockClient(mockCtrl)
 	p.API = mockAPI
 	p.Store = mockedStore
+	p.Client = mockedClient
```
Comment (ayusht2810) [inline server/plugin/command_test.go:33]: can't we use the function `setupMockPlugin` here?

File: server/plugin/taskLinkPreview_test.go [hunk 23]
```diff
@@ -56,3 +56,26 @@ func TestPostPullRequestPreview(t *testing.T) {
 		})
 	}
 }
+
+func TestPostBuildDetailsPreview(t *testing.T) {
+	p := Plugin{}
+	mockCtrl := gomock.NewController(t)
+	mockedClient := mocks.NewMockClient(mockCtrl)
+	p.Client = mockedClient
```
Comment (ayusht2810) [inline server/plugin/taskLinkPreview_test.go:64]: same

File: server/plugin/taskLinkPreview_test.go [hunk 24]
```diff
@@ -56,3 +56,26 @@ func TestPostPullRequestPreview(t *testing.T) {
...
+	for _, testCase := range []struct {
+		description string
+		linkData    []string
+	}{
+		{
+			description: "PostBuildDetailsPreview: valid",
+			linkData:    []string{"https:", "", "dev.azure.com", "abc", "xyz", "_build", "results?buildId=50&view=results"},
```
Comment (ayusht2810) [inline server/plugin/taskLinkPreview_test.go:71]:
```suggestion
			linkData:    []string{"https:", "", "test.com", "abc", "xyz", "_build", "results?buildId=50&view=results"},
```

Comment (ayusht2810) [inline server/plugin/taskLinkPreview_test.go:71]:
```suggestion
			linkData:    []string{"https:", "", "test.com", "abc", "xyz", "_build", "results?buildId=50&view=results"},
```

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api_test.go [hunk 25]
```diff
@@ -266,21 +269,44 @@ func TestHandleLink(t *testing.T) {
...
-				mockedStore.EXPECT().StoreProject(&serializers.ProjectDetails{
-					MattermostUserID: "mockMattermostUserID",
-					ProjectName:      "Mockproject",
-					OrganizationName: "mockorganization",
-				}).Return(nil)
+				if !testCase.isProjectLinked {
+					mockedClient.EXPECT().Link(gomock.Any(), gomock.Any()).Return(&serializers.Project{}, testCase.statusCode, testCase.err)
+					mockedClient.EXPECT().CheckIfUserIsProjectAdmin(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(http.StatusOK, nil)
+					mockedStore.EXPECT().StoreProject(&serializers.ProjectDetails{
+						MattermostUserID: "mockMattermostUserID",
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:305]: you can create a test util function here we will return you mock projectDetails and use that everywhere

File: server/plugin/api_test.go [hunk 26]
```diff
@@ -960,3 +1070,102 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+			description: "HandleGetSubscriptionFilterPossibleValues: Error fetching subscription filter possible values",
+			body: `{
+				"organization": "mockOrganization",
+				"projectId": "mockProjectID",
+				"eventType": "mockEventType",
+				"repositoryId": "mockRepositoryID",
+				"filters": ["mockFilter1", "mockFilter2"]
+				}`,
+			statusCode:                             http.StatusInternalServerError,
+			getSubscriptionFilterPossibleValuesErr: errors.New("failed to fetch subscription filters possible values"),
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:1138]: Check this everywhere
```suggestion
			getSubscriptionFilterPossibleValuesErr: errors.New("failed to fetch the subscription filters possible values"),
```

File: server/plugin/client.go [hunk 27]
```diff
@@ -268,7 +268,7 @@ func (c *client) CheckIfUserIsProjectAdmin(organizationName, projectID, pluginUR
 
 	_, statusCode, err := c.CallJSON(c.plugin.getConfiguration().AzureDevopsAPIBaseURL, subscriptionURL, http.MethodPost, mattermostUserID, payload, nil, nil)
 	if err != nil {
-		return statusCode, errors.Wrap(err, "failed to create subscription")
+		return statusCode, errors.Wrap(err, "failed to check if user is project admin")
```
Comment (Kshitij-Katiyar) [inline server/plugin/client.go:271]:
```suggestion
		return statusCode, errors.Wrap(err, "failed to check if user is a project admin")
```

File: server/plugin/client_test.go [hunk 28]
```diff
@@ -317,6 +350,90 @@ func TestCall(t *testing.T) {
...
+	}{
+		{
+			description: "CheckIfUserIsProjectAdmin: valid",
+			statusCode:  http.StatusOK,
+		},
+		{
+			description:   "CheckIfUserIsProjectAdmin: with error",
+			err:           errors.New("failed to check user permissions"),
+			statusCode:    http.StatusInternalServerError,
+			expectedError: "failed to check if user is project admin: failed to check user permissions",
```
Comment (Kshitij-Katiyar) [inline server/plugin/client_test.go:370]:
```suggestion
			expectedError: "failed to check if the user is a project admin: failed to check user permissions",
```

File: server/plugin/command_test.go [hunk 29]
```diff
@@ -63,31 +73,178 @@ func TestExecuteCommand(t *testing.T) {
...
+		{
+			description: "ExecuteCommand: repos add subscription command",
+			isConnected: true,
+			commandArgs: &model.CommandArgs{Command: "/azuredevops repos subscription add"},
+		},
+		{
+			description:     "ExecuteCommand: repos delete subscription command",
+			isConnected:     true,
+			commandArgs:     &model.CommandArgs{Command: "/azuredevops repos subscription delete mockSubscriptionID"},
+			isDeleteCommand: true,
```
Comment (Kshitij-Katiyar) [inline server/plugin/command_test.go:157]: i think there is no need for these kind of variables here you can get them from the arguments of the command and same for service type. See if this is possible

File: server/plugin/taskLinkPreview_test.go [hunk 30]
```diff
@@ -56,3 +54,25 @@ func TestPostPullRequestPreview(t *testing.T) {
...
+func TestPostBuildDetailsPreview(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	mockedClient := mocks.NewMockClient(mockCtrl)
+	p := setupMockPlugin(nil, nil, mockedClient)
+	for _, testCase := range []struct {
+		description string
+		linkData    []string
+	}{
+		{
+			description: "PostBuildDetailsPreview: valid",
```
Comment (Kshitij-Katiyar) [inline server/plugin/taskLinkPreview_test.go:67]: dont use loops if there is just one subtest case

File: server/plugin/command_test.go (hunk 29 above)
Comment (raghavaggarwal2308) [reply server/plugin/command_test.go:157]: this will be a little hard to implement. We can discuss this over a call

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/command_test.go (hunk 29 above)
Comment (raghavaggarwal2308) [reply server/plugin/command_test.go:157]: discussed

**Pull request merged**